File: src/main/kotlin/fun/basic/model/function/ParametricTransformation.java
package fun.basic.model.function;

import fun.basic.model.domain.Domain;
import fun.basic.model.domain.GenericDomain;
import fun.basic.model.domain.Interval;
import fun.basic.model.domain.Vector;

import java.util.HashSet;
import java.util.Set;

public interface ParametricTransformation<CoDomPoint extends Vector<CoDomPoint>> {

    /**
     * transform a point from one Domain to another.
     *
     * @param point the point to transform
     * @return the image of the point
     */
    CoDomPoint transform(double point);

    /**
     * transform a Domain to another.
     *
     * @param domain thedomain to transform
     * @return the image of the domain
     */
    default Domain<CoDomPoint> transform(Set<Double> domain) {
        Set<CoDomPoint> points = new HashSet<CoDomPoint>();
        for (double point : domain) {
            points.add(transform(point));
        }
        return new GenericDomain<CoDomPoint>(points);
    }

    /**
     * transform a Domain to another.
     *
     * @param domain thedomain to transform
     * @return the image of the domain
     */
    default Domain<CoDomPoint> transform(Interval domain) {
        return transform(domain.getPoints());
    }
}

File: src/main/kotlin/fun/basic/model/function/ParametricTransformation.kt
package `fun`.basic.model.function

import `fun`.basic.model.domain.Domain
import `fun`.basic.model.domain.GenericDomain
import `fun`.basic.model.domain.Interval
import `fun`.basic.model.domain.Vector

interface ParametricTransformation<CoDomPoint: Vector<CoDomPoint>> {

    /**
     * transform a point from one Domain to another.
     *
     * @param point the point to transform
     * @return the image of the point
     */
    fun transform(point: Double): CoDomPoint

    /**
     * transform a Domain to another.
     *
     * @param domain thedomain to transform
     * @return the image of the domain
     */
    fun transform(domain: Set<Double>): Domain<CoDomPoint> {
        val points: MutableSet<CoDomPoint> = HashSet()
        for (point in domain) {
            points.add(transform(point))
        }
        return GenericDomain(points)
    }

    /**
     * transform a Domain to another.
     *
     * @param domain thedomain to transform
     * @return the image of the domain
     */
    fun transform(interval: Interval): Domain<CoDomPoint> {
        return transform(interval.points)
    }
}